fdas/main.go

```diff
@@ -0,0 +1,299 @@
+package main
+
+import (
+	"fmt"
+	"net/http"
+	"os"
+	"strings"
+
+	"github.com/gin-gonic/gin"
+	"github.com/google/uuid"
+	capi "github.com/hashicorp/consul/api"
+	"github.com/syndtr/goleveldb/leveldb"
+	"gopkg.in/yaml.v3"
+)
+
+// Structs
+
+type appConfig struct {
+	LevelDBLocation string `yaml:"LevelDBLocation"`
+}
+
+type fdasRelations struct {
+	Union []string `json:"union"`
+}
+
+type fdasConfig struct {
+	Namespace string                    `json:"namespace"`
+	Relations map[string]*fdasRelations `json:"relations"`
+}
+
+type fdasACL struct {
+	Object   string `json:"object"`
+	Relation string `json:"relation"`
+	User     string `json:"user"`
+}
+
+type Set struct {
+	list map[string]struct{}
+}
+
+func (s *Set) Has(v string) bool {
+	_, ok := s.list[v]
+	return ok
+}
+
+func (s *Set) Add(v string) {
+	s.list[v] = struct{}{}
+}
+
+func (s *Set) Remove(v string) {
+	delete(s.list, v)
+}
+
+func (s *Set) Clear() {
+	s.list = make(map[string]struct{})
+}
+
+func (s *Set) Size() int {
+	return len(s.list)
+}
+
+func NewSet() *Set {
+	s := &Set{}
+	s.list = make(map[string]struct{})
+	return s
+}
+
+func (s *Set) AddMulti(list ...string) {
+	for _, v := range list {
+		s.Add(v)
+	}
+}
+
+// Endpoints
+
+func postACL(tokens *Set, db *leveldb.DB, kv *capi.KV) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		if !tokens.Has(c.Query("token")) {
```

Comment:
Can we use Header data, instead of Query params (they are encrypted in the https, but they will be stored in browser history and can be accidentally shared)

Reply:
Idea for Header data path parameter is good I will make necessary changes.

Also as FastDistributedAuthorizationSystem will be used as SaaS for auth probably in some developers code I don't think we will encounter problem of browser caching and leaking token. But to make sure I will add endpoint to invalidate token 👍